### ISSUE TITLE
Specify type explicitelly should work for `_`

### ISSUE THREAD

**Issue by matklad** [rust-analyzer/rust-analyzer#1965]
See: https://github.com/rust-analyzer/rust-analyzer/issues/1580#issuecomment-538717634

TL;DR: in 

```rust
let xs: Vec<_> = vec![1, 2, 3]
```

the assist should fill `_` as `i32`.

We already has an `add_explicit_type` assist here:

https://github.com/rust-analyzer/rust-analyzer/blob/c73193efe29ff4072fd19471fb22cdd2ee05556c/crates/ra_assists/src/assists/add_explicit_type.rs

It should be enhanced to support diffing specified and actual type and replacing `_` with the type. 

Rough steps to implment:

* add `type_of_typeref` method for `SourceAnalyzer`, that turns `ast::TypeRef` into `hir::Ty`
* write code that walks two `hir::Ty`s in lockstep and specifies replacements for all `_`, or finds that this is impossible
* make assist available if cursor is on `_`
* write the tests (in the same file)

**Comment by mattyhall** [comment]
Hi. I was looking into this but am a bit confused. I'm not sure what the value of the first step is ("add type_of_typeref"). Looking at the hover code it looks like it should be possible to get [the type from the pattern itself](https://github.com/rust-analyzer/rust-analyzer/blob/d3872964f88a0d751c428c150bb40d8b4f4c89a9/crates/ra_ide_api/src/hover.rs#L250). However, using this code in add_explicit_type doesn't seem to work. Does the SourceAnalyser in an assist have less information than one in the hover code?

**Comment by matklad** [comment]
> Looking at the hover code it looks like it should be possible to get the type from the pattern itself

Right, yeah, I think we should just use that!

>Does the SourceAnalyser in an assist have less information than one in the hover code?

Nope, it should work.

**Comment by lnicola** [comment]
Fixed in #2948 AFAICT?
